### DISCUSSION TITLE
How to debug why I don't see errors?

### DISCUSSION THREAD

**Discussion by yanshay** [rust-lang/rust-analyzer#17881] [Q&A] [answered]
I'm developing for `esp32` on `lazyvim` (which in turn uses `rustaceanvim` plugin for `neovim`).

I don't get compilation errors. I believe this is due to cargo check.

On the command-line in order to get cargo check to work I need to write:
`cargo +esp check --release --target xtensa-esp32-none-elf
`
So I configured lazyvim as below which I presume eventually arrives to rust-analyzer, but it doesn't help.

I exhausted everything I could think of and could find online.
How can I debug this? Is there a way to see what command line eventually cargo check gets to run?

```
      default_settings = {
        -- rust-analyzer language server configuration
        ["rust-analyzer"] = {
          cargo = {
            allFeatures = false,                    <----------------- changed this
            allTargets = false,                       <----------------- changes this
            loadOutDirsFromCheck = true,
            buildScripts = {
              enable = true,
            },
            extraArgs = {"+esp", "--release"},  <------------- added this
            targets = {"xtensa-esp32-none-elf"}. <---------- added this
          },
          check = {
            extraArgs = {"+esp", "--release"}, <------------- tried also this
          },
          -- Add clippy lints for Rust.
          checkOnSave = true,
          procMacro = {
            enable = true,
            ignored = {
              ["async-trait"] = { "async_trait" },
              ["napi-derive"] = { "napi" },
              ["async-recursion"] = { "async_recursion" },
            },
          },
        },
```

**Comment by Veykril** [comment]
Do the logs say anything?

**Reply by yanshay** [reply]
The only logs I'm aware of are the `lsp.log` (in `.local/state/nvim`). See below.

Are there other logs I can see (or activate to see)? 

I think it's about the command line parameters that are passed to cargo but I don't see how I can see them.
I don't see the below is correlated with the issues I have, but at the same time I don't understand what they do mean.

```
[START][2024-08-13 19:25:09] LSP logging initiated
[ERROR][2024-08-13 19:25:09] .../vim/lsp/rpc.lua:770    "rpc"   "/Users/my_user/.local/share/nvim/mason/bin/taplo"   "stderr"        ' INFO taplo: registered request handler method="textDocument/foldingRange"\n INFO taplo: registered request handler method="initialize"\n INFO taplo: registered request handler method="textDocument/documentSymbol"\n INFO taplo: registered request handler method="textDocument/formatting"\n INFO taplo: registered request handler method="textDocument/completion"\n INFO taplo: registered request handler method="textDocument/hover"\n INFO taplo: registered request handler method="textDocument/documentLink"\n INFO taplo: registered request handler method="textDocument/semanticTokens/full"\n INFO taplo: registered request handler method="textDocument/prepareRename"\n INFO taplo: registered request handler method="textDocument/rename"\n INFO taplo: registered notification handler method="initialized"\n INFO taplo: registered notification handler method="textDocument/didOpen"\n INFO taplo: registered notification handler method="textDocument/didChange"\n INFO taplo: registered notification handler method="textDocument/didSave"\n INFO taplo: registered notification handler method="textDocument/didClose"\n INFO taplo: registered notification handler method="workspace/didChangeWorkspaceFolders"\n INFO taplo: registered notification handler method="workspace/didChangeConfiguration"\n INFO taplo: registered request handler method="taplo/convertToToml"\n INFO taplo: registered request handler method="taplo/listSchemas"\n INFO taplo: registered request handler method="taplo/associatedSchema"\n INFO taplo: registered request handler method="taplo/convertToJson"\n INFO taplo: registered notification handler method="taplo/associateSchema"\n INFO taplo: LSP server listening transport="stdio"\n'
[ERROR][2024-08-13 23:55:51] .../vim/lsp/rpc.lua:770    "rpc"   "/Users/my_user/.local/share/nvim/mason/bin/taplo"   "stderr"        "ERROR taplo: no request handler registered method=workspace/willRenameFiles\n"
[ERROR][2024-08-13 23:55:51] .../vim/lsp/rpc.lua:770    "rpc"   "/Users/my_user/.local/share/nvim/mason/bin/taplo"   "stderr"        " WARN taplo: no notification handler registered method=workspace/didRenameFiles\n"
[ERROR][2024-08-13 23:56:00] .../vim/lsp/rpc.lua:770    "rpc"   "/Users/my_user/.local/share/nvim/mason/bin/taplo"   "stderr"        "ERROR taplo: no request handler registered method=workspace/willRenameFiles\n"
[ERROR][2024-08-13 23:56:00] .../vim/lsp/rpc.lua:770    "rpc"   "/Users/my_user/.local/share/nvim/mason/bin/taplo"   "stderr"        " WARN taplo: no notification handler registered method=workspace/didRenameFiles\n"
```

**Reply by Veykril** [reply]
Those are `taplo` logs, not `rust-analyzer`, is `rust-analyzer` even running for you? (assuming these are the only logs you can find)

**Reply by yanshay** [reply]
I'm not an expert with this, here is what I know:
1. When I perform various operations (open a rust file, saving, etc.) I see in Neovim messages stating rust-analyzer is running, see below screen capture. When I save for example I can see` (cargo check) rust-analyzer`for a short time. Sometimes even two lines like that one on top of another. I assume this means rust-analyzer is running.
2. The only logs I know of are `lsp.log` file. In the past I've seen there rust-analyzer logs when there were rust-analyzer issues (like when dependencies failed to compile that weren't reported in Neovim, and in rust-analyzer repo I received an answer that rust-analyzer 'eats' those issues). Now I don't see there rust-analyzer issues. Maybe there are others logs, if you direct me I can look there.
3. I don't see in Neovim errors that appear in cargo check.
4. This his happens only on esp32 projects, errors I can see in non esp32 projects in neovim I don't see in esp32 projects. So I think it has to do with the command line params passed to rust-analyzer. I configured everything to the best of my knowledge, in ways that worked in the past, but now they don't work and I'm looking for ways to debug this.
![image](https://github.com/user-attachments/assets/685115e8-c769-4995-b5e3-e5a7159dcb45)

**Comment by yanshay** [comment]
BTW - What's interesting is that now I see I do get errors, just different errors in Neovim than I get in cargo check, maybe that can help troubleshoot the issue. It states that "target_arch" is disabled, but I have set the cargo.targets in my config. 

In Neovim I see:

![image](https://github.com/user-attachments/assets/c60e7c63-e4bf-459d-be58-1a5cca3a882e)

Cargo check shows (the proper errors):
```
error: 2 positional arguments in format string, but there is 1 argument
   --> src/main.rs:103:19
    |
103 |         println!("{}, {}", "Hello, world!");
    |                   ^^  ^^   ---------------

error[E0106]: missing lifetime specifier
  --> src/my_mqtt.rs:26:31
   |
26 |     tls: AsyncConnectedSession<a, T, BUFFER_SIZE>,
   |                               ^ expected named lifetime parameter
   |
help: consider using the `'a` lifetime
   |
26 |     tls: AsyncConnectedSession<'a, a, T, BUFFER_SIZE>,
   |                                +++

error[E0412]: cannot find type `a` in this scope
  --> src/my_mqtt.rs:26:32
   |
22 | pub struct MyMqtt<'a, T, const BUFFER_SIZE: usize>
   |                       - similarly named type parameter `T` defined here
...
26 |     tls: AsyncConnectedSession<a, T, BUFFER_SIZE>,
   |                                ^
   |
help: a type parameter with a similar name exists
   |
26 |     tls: AsyncConnectedSession<T, T, BUFFER_SIZE>,
   |                                ~
help: you might be missing a type parameter
   |
22 | pub struct MyMqtt<'a, T, const BUFFER_SIZE: usize, a>
   |                                                  +++

error[E0107]: struct takes 2 generic arguments but 3 generic arguments were supplied
   --> src/my_mqtt.rs:26:10
    |
26  |     tls: AsyncConnectedSession<a, T, BUFFER_SIZE>,
    |          ^^^^^^^^^^^^^^^^^^^^^       ----------- help: remove this generic argument
    |          |
    |          expected 2 generic arguments
    |
note: struct defined here, with 2 generic parameters: `T`, `BUFFER_SIZE`
   --> /Users/*****/OpenSourceProj/esp-mbedtls/esp-mbedtls/src/lib.rs:985:16
    |
985 |     pub struct AsyncConnectedSession<'a, T, const BUFFER_SIZE: usize>
    |                ^^^^^^^^^^^^^^^^^^^^^     -  ------------------------

Some errors have detailed explanations: E0106, E0107, E0412.
For more information about an error, try `rustc --explain E0106`.
error: could not compile `nfc-filament` (bin "nfc-filament") due to 4 previous errors

```

**Reply by Veykril** [reply]
Those `No proc-macros present for crate` errors point to something more problematic, hence why I am confused that you don't get any error logs as those usually indicate a problem with running build scripts (and hence also most other cargo commands)

**Reply by Veykril** [reply]
Can you set `RA_LOG="warn,flycheck=debug,rust_analyzer::reload=info,project_model=info"` for your session and see if that gives you any useful logs?

**Reply by yanshay** [reply]
To do that in the environment session of neovim?

**Reply by yanshay** [reply]
I set this as env variable (assuming that's what you meant), in lsp.log I don't see any new logs coming out.

**Reply by Veykril** [reply]
Yes in the session for your lsp client. If you still dont get logs I am very confused as we should be definitely logging things now. Might be that your lsp client has some additional logging settings? I don't know vim unfortunately

**Reply by yanshay** [reply]
I've made some progress finding the log file, it's location is shown by the command LspInfo in Neovim.
So maybe now you would be able to help? 

I think I found the root cause but don't understand how to fix.
To compile the project I normally use `cargo +esp --release --target xtensa-exp32-none-elf`.
The `+esp` has to come first on the command line!

So I added to the configuration:
```
      default_settings = {
        -- rust-analyzer language server configuration
        ["rust-analyzer"] = {
          server = {
            -- extraEnv = { {["RA_LOG"]="project_model=debug"} },
            extraEnv = { ["RA_LOG"]="project_model=debug" },
          },
          cargo = {
            allFeatures = false,
            loadOutDirsFromCheck = true,
            buildScripts = {
              enable = true,
            },
            extraArgs = {"+esp", "--release"},       <-------------- added
            allTargets = false,                                  <-------------- added
            target = "xtensa-esp32-none-elf".      <-------------- added
          },
          -- Add clippy lints for Rust.
          checkOnSave = true,
          procMacro = {
            enable = true,
            ignored = {
              ["async-trait"] = { "async_trait" },
              ["napi-derive"] = { "napi" },
              ["async-recursion"] = { "async_recursion" },
            },
          },
        },
```
I attached the rust-analyzer log below as `ra2.log` .

The main thing I see is the `error: unexpected argument '+esp' found` - so the +esp reached there, but it is not understood by cargo, probably because it is place in the wrong position on the cargo command line? I also see `--release` is there.
 
The error `error: Error loading target specification: Could not find specification for target "xtensa-esp32-none-elf". Run `rustc --print target-list` for a list of built-in targets` is there because the `+esp` is missing.

- Is there a way to get that +esp to appear earlier on the cargo command line? or to get cargo to understand it needs to use the esp rustc?


Here is the full log:
[ra2.log](https://github.com/user-attachments/files/16628033/ra2.log)

**Reply by Veykril** [reply]
Oh the esp is a toolchain override isnt it? You could try setting the `RUSTUP_TOOLCHAIN` env var to `esp` in the `server.extraEnv`, I *think* that should cause everything in r-a to use the appropriate toolchain then. (so you can drop the `+esp` in the command configs. Otherwise you will have to use the `overrideCommand` configs and fully overwrite the invoked commands, as the `extraArgs` are always appended to the command invocation.

**Reply by yanshay** [reply]
I tried the `RUSTUP_TOOLCHAIN`, it didn't work because then I got an error that the target is not known. 
 So eventually I linked the rust-analyzer from the esp-toolchain to the stable. I think with that only the rust-analyzer is from the stable toolchain and all others (cargo, etc.) are from the esp toolchain.

I thought about overrideCommand, I actually tried to put there 'cargo +esp' but seems that doesn't work, need to write an alternative command, and the link was a simpler option. I documented it all in the next message for others.

**Comment by yanshay** [comment, marked as answer]
I found a way to get this working, not sure if the best/only way but documenting for others who may stumble on this:
1. set rust-toolchain.toml to
```
[toolchain]
channel = "esp" # This is the important line
targets = ["xtensa-esp32-none-elf"] # This does not affect the rust-analyzer but for build it works
```
2. ln -sf /Users/user/.rustup/toolchains/stable-x86_64-apple-darwin/bin/rust-analyzer ~/.rustup/toolchains/esp/bin/rust-analyzer
This is so with esp there will still be rust-analyzer found.
The other options I saw elsewhere to set RUST_TOOLCHAIN didn't work for me. Other things broke. Maybe possible to get it to work.

Found it here: https://www.reddit.com/r/rust/comments/13d2tls/rustanalyzer_and_toolchain_for_esp/

3. In LazyVim
```
return {
  "mrcjkb/rustaceanvim",
  opts = {
    server = {
      default_settings = {
        -- rust-analyzer language server configuration
        ["rust-analyzer"] = {
          server = {
            -- extraEnv = { {["RA_LOG"]="project_model=debug"} }, -- For debugging rust-analyzer, to see log location do :LspInfo in neovim
          },
          cargo = {
            allFeatures = false, -- important
            extraArgs = {"--release"}, -- probably not required, but better since used for building
            allTargets = false,  -- Not required
            -- target = "xtensa-esp32-none-elf", -- can be avoided if using .cargo/config.toml as described below. CAN'T be a list even if in rust-analyzer doc it says otherwise
          },
        },
      },
    },
  },
}
```


I think that's it.
